src/middlewared/middlewared/utils/io.py

```diff
@@ -1,24 +1,124 @@
 import os
+import enum
+import stat
+from tempfile import NamedTemporaryFile
 
 
-def write_if_changed(path, data):
+ID_MAX = 2 ** 32 - 2
+
+
+class FileChanges(enum.IntFlag):
+    CONTENTS = enum.auto()
+    UID = enum.auto()
+    GID = enum.auto()
+    PERMS = enum.auto()
+
+
+class UnexpectedFileChange(Exception):
+    def __init__(self, path, changes):
+        self.changes = changes
+        self.path = path
+        self.changes_str = ', '.join([x.name for x in FileChanges if self.changes & x])
+
+    def __str__(self):
+        return f'{self.path}: unexpected change in the following file attributes: {self.changes_str}'
+
+
+def write_if_changed(path, data, uid=0, gid=0, perms=0o755, dirfd=None, raise_error=False):
+    """
+    Commit changes to a configuration file.
+    `path` - path to configuration file. May be relative to a specified `dirfd`
+
+    `data` - expected file contents. May be bytes or string
+
+    `uid` - expected numeric UID for file
+
+    `gid` - expected numeric GID for file
+
+    `perms` - numeric permissions that file should have
+
+    `raise_error` - raise an exception if file ownership or permissions have unexpectedly
+    changed.
+    """
 
     if isinstance(data, str):
         data = data.encode()
 
-    if isinstance(path, int):
-        path = f'/proc/self/fd/{path}'
+    if not isinstance(perms, int):
+        raise ValueError('perms must be an integer')
+
+    if perms & ~(stat.S_IRWXU | stat.S_IRWXG | stat.S_IRWXO):
+        raise ValueError(f'{perms}: invalid mode')
```

Comment:
Why is it invalid?

Reply:
Okay. I'll specify that we only allow setting RWX on user, group, other.